onnx/defs/tensor/defs.cc

```diff
@@ -2426,11 +2435,13 @@ ONNX_OPERATOR_SET_SCHEMA(
         .Input(
             1,
             "grid",
-            "Input offset, 4-D tensor of shape (N, H_out, W_out, 2), "
-            "where H_out and W_out are the height and width of grid and output, "
-            "Grid specifies the sampling pixel locations normalized by the input spatial dimensions. "
-            "Therefore, it should have most values in the range of [-1, 1]. "
-            "If grid has values outside the range of [-1, 1], the corresponding outputs will be handled as defined by padding_mode.",
+            "Input offset of shape (N, D1_out, D2_out, …, Dr_out, r), where D1_out, D2_out, …, "
```

Comment:
If I understand correctly, the dimensions (N, D1_out, ... Dr_out) are irrelevant to the internal computation. That is, one could equivalently flatten these dimensions out. Is that right?  This particular format may reflect how the callers use it, but doesn't have any impact on the internal implementation of the op.

Is this right? If so, I wonder if it would be simpler to say that the dimensions (D1_out, ... Dr_out) can be of any length (not necessarily length r)? Just trying to understand this better. I realize this generalization may not have much use in practice.

Reply:
You are correct. For the grid, it's just a bunch of points in the r-dimensional space that's waiting to be interpolated from the r-dimensional space in the input. PyTorch enforces this r-dimension to r-dimension mapping, and we also follow their convention. If the user wants to do some mappings such as r1-dimension to r2-dimension mapping, they can do grid sample followed by reshaping the output tensor.

Reply:
Enforcing r-dimension to r-dimension mapping allows us to call the GridSample spatial, volumetric, etc.